README.md
![readme banner](./assets/banner.png)

<div align="center">
  <h1>Better GitHub</h1>
  <p><strong>🎨 Enhance your code reading experience on GitHub</strong></p>
  <h3 align="center">
    <a href="https://addons.mozilla.org/en-US/firefox/addon/bettergithub/">Firefox</a>
    <span> · </span>
    <a href="https://chrome.google.com/webstore/detail/better-github/ammeaejgjdeifjekkofhnliedhccbmgp">Chrome</a>
  </h3>
</div>

# Demo

<img src="./assets/demo.gif">

# Get the extension

### [Firefox](https://addons.mozilla.org/en-US/firefox/addon/bettergithub/)

### [Chrome](https://chrome.google.com/webstore/detail/better-github/ammeaejgjdeifjekkofhnliedhccbmgp)

# Features

### Simple interface to use the extension

![popup](./assets/popup.png)

`Better Github` allows you to apply custom fonts and modify font size of code text. These styles are applied to :

> Example below have the following Better GitHub's config: `Font Name` is [Hack](https://github.com/source-foundry/Hack) and `Font Size` is **14**.

- All the text inside a file while viewing( reading ) it.

   <img src="./assets/feature-1-example.png">

- Code in README files that are inside "`<code>`" blocks also known as `<pre>` tags.

  <img src="./assets/feature-2-example.png">

- Code in pull request diffs.

  <img src="./assets/feature-3-example.png">

**Note:** If the custom styles are not applied( it can happen sometimes ), refresh the page.

# Motivation

Default font size of code text was very small which gave me had a hard time reading code in any repository and in PR diffs. I also wanted my code editor font and GitHub font to be same.

Couldn't find anything existing to help me solve my problem, so I created this simple yet powerful extension for Chrome browser. If something does exist though, let me know, alright?

I called it `Better GitHub` inspired by `Better Twitch TV` and `Better Discord`.

# Todos

> Context: I created this list of tasks after I decided to make this extension public. Initially the code was very small and the font styles were hard coded. If I needed to modify the styles, I had to change it in the source code, reload the extension and refresh GitHub pages to reflect the updates. Consider this list as the roadmap for the project.

- [x] Instead of hard coding the `fontFamily` and `fontSize`, we should be able to allow the user to choose a font size and font family of their choice (which they have installed on their machine).
- [x] Create a browser action popup to show an interface to allow the user to customise their extension's settings for `fontFamily` and `fontSize`.
- [x] Integrate browser action popup with chrome API to persist and read user's settings for `fontFamily` and `fontSize` from and to the storage.
- [x] Apply styles without reloading whenever the settings are changed from the browser action popup interface.
- [x] If no custom font styles are set, load GitHub's default font styles.
- [x] User can reset font styles to GitHub's default font styles.
- [x] Add GIF to show the usage( demo ) of the extension.
- [x] Installation instructions on how to clone/download this repository, install the extension and use it.
- [x] Logo for the extension that will be used as favicon, icon, in documentation etc.
- [x] Firefox support as it was requested by [others](https://dev.to/ceoshikhar/enhance-your-code-reading-experience-on-github-with-this-chrome-extension-24ei).
- [x] Publish it as an official Mozilla Firefox addon.
- [x] Publish it as an official Chrome extension on Chrome Web Store.
- [x] Allow user to change only one property instead of all properties being mandatory.
- [ ] Add `CONTRIBUTING.md` to help others so that they can contribute to the project.


CHANGELOG.md
# Changelog

# Unversioned

### Fixes

- Some font names don't apply, for example 'Terminus (TTF) for Windows'. [PR #14](https://github.com/ceoshikhar/better-github/pull/14).


# 1.2.0 (16-01-2022)

### Features

- Apply styles to `code` HTML elements. [PR #13](https://github.com/ceoshikhar/better-github/pull/13).

# 1.1.0 (16-01-2022)

### Features

- Can provide just one style property and it will apply that style property instead of making the user provide values for all the style properties. [PR #12](https://github.com/ceoshikhar/better-github/pull/12).

# 1.0.3 (26-12-2021)

### Features

- Can customize line height.

# 1.0.2 (07-07-2021)

### Fixes

- Mouse cursor not falling in correct place while editing a file. [Issue #6](https://github.com/ceoshikhar/better-github/issues/6) - Fixed by [PR #10](https://github.com/ceoshikhar/better-github/pull/10).

- Stopped working on github gists. [Issue #8](https://github.com/ceoshikhar/better-github/issues/8) - Fixed by [PR #9](https://github.com/ceoshikhar/better-github/pull/9). Thanks to [Aahnik Daw](https://github.com/aahnik).

# 1.0.1 (01-04-2021)

### Features

- Support for Firefox. Released it as an official mozilla firefox [addon](https://addons.mozilla.org/en-US/firefox/addon/bettergithub)
- Scripts to build/package the extension( zip ) and generate manifest.json

### Fixes and refactors

- Extra scroll( overflow ) on `<body>` in Popup on Firefox
- Refactor Popup CSS styles to `styles.css`

# 1.0.0 (29-03-2021)

First public release


better-github.js
// Store the current set font styles to this object so that we don't have to
// fetch them from chrome storage API again and again.
// This improves the performance as we don't do the async chrome storage API
// read calls every time `applyCurrentSetStyles` is called.
const cache = {
    dirty: false,
    fontName: null,
    fontSize: null,
    lineHeight: null,
};

// These styles were taken from Google Chrome's(Windows) inspect element tool
// on 24/3/2021. These default styles might change but it's not really that
// important otherwise it would defeat the whole purpose of `Better Github`.
const DEFAULT_FONT_FAMILY =
    "SFMono-Regular,Consolas,Liberation Mono,Menlo,monospace";
const DEFAULT_FONT_SIZE = "12px";
const DEFAULT_LINE_HEIGHT = 1.4;

// Do the magic once the DOM is completely loaded.
window.onload = init();

// Entry point of the extension to interact with the DOM on GitHub.
function init() {
    // Apply font styles when DOM tree is completely loaded.
    applyCurrentSetStyles();

    // Apply font styles when DOM tree is mutated.
    reApplyStylesOnDOMChange();
}

// Apply the currently set font styles if they exist
async function applyCurrentSetStyles() {
    if (isUserEditingFile()) return;

    const currentSetFontStyles = await getCurrentSetFontStyles();
    if (!currentSetFontStyles) return;

    const { fontFamily, fontSize, lineHeight } = currentSetFontStyles;
    applyStyles(fontFamily, fontSize, lineHeight);
}

// Apply the custom styles whenever something on the DOM changes. Following are
// some scenarios why this is important.
//
// 1. With larger PRs, sometimes diffs are loaded lazily. Which means, a DOM
// mutation is happening. If we don't re-apply the styles, the newly loaded
// code text in the diff will have the defaul styles.
//
// 2. GitHub is an SPA. Which means, DOM is loaded only once. When you navigate
// to other "pages" on GitHub, the DOM is changing( mutation ) and if we don't
// re-apply the styles, the code text on new page will have default styles.
//
// We don't care to check what the change is happening on the DOM, we just
// re-apply the styles no matter what's happening. This shouldn't be affecting
// the performance as all we are doing is changing the CSS styles, also there
// are not so many DOM changes happening once the entire page content is loaded.
function reApplyStylesOnDOMChange() {
    // We pass `applyCurrentSetStyles` as the callback to `Mutation Observer`.
    // This means, whenever a DOM mutation is observed it fires the callback.
    const observer = new MutationObserver(applyCurrentSetStyles);

    observer.observe(document, { childList: true, subtree: true });
}

// Reset the font styles to Github's default.
function resetStyles() {
    // Clear the storage so that when we reload/refresh or visit GitHub in another
    // tab, we don't apply any custom styles including the above two mentioned.
    // The real(coming from GitHub) GitHub's default styles are used.
    clearStorage();

    // We apply the GitHub's "default styles" so that we don't have to manually
    // refresh the current open tabs and this makes it feel "reactive".
    applyStyles(DEFAULT_FONT_FAMILY, DEFAULT_FONT_SIZE, DEFAULT_LINE_HEIGHT);
}

// This is where the magic happens. As the name suggests, the function is
// responsible for applying the font styles to the code text elements on a page.
//
// Apply font styles (font-name & font-size).
// These styles are applied to :
// - All the text inside a file while viewing( reading ) it.
// - Code in README files that are inside "`<code>`" blocks also known as `<pre>` tags.
// - Code in pull request diffs.
function applyStyles(fontFamily, fontSize, lineHeight) {
    const codeTextElements = document.getElementsByClassName("blob-code-inner");
    const codeLineNumElements = document.getElementsByClassName("blob-num");
    const preElements = document.querySelectorAll("pre");
    const codeElements = document.querySelectorAll("code");

    for (let i = 0; i < codeTextElements.length; i++) {
        applyStyle(codeTextElements[i], "fontFamily", fontFamily);
        applyStyle(codeTextElements[i], "fontSize", fontSize);
        applyStyle(codeTextElements[i], "lineHeight", lineHeight);
    }

    for (let i = 0; i < codeLineNumElements.length; i++) {
        applyStyle(codeLineNumElements[i], "fontFamily", fontFamily);
        applyStyle(codeLineNumElements[i], "fontSize", fontSize);
        applyStyle(codeLineNumElements[i], "fontHeight", lineHeight);
    }

    for (let i = 0; i < preElements.length; i++) {
        applyStyle(preElements[i], "fontFamily", fontFamily);
        applyStyle(preElements[i], "fontSize", fontSize);
        applyStyle(preElements[i], "fontHeight", lineHeight);
    }

    for (let i = 0; i < codeElements.length; i++) {
        applyStyle(codeElements[i], "fontFamily", fontFamily);
        applyStyle(codeElements[i], "fontSize", fontSize);
        applyStyle(codeElements[i], "fontHeight", lineHeight);
    }
}

/**
 * @param {HTMLElement} el The element to apply the style to.
 * @param {string} prop The style property to change.
 * @param {string} value The value to apply for the given `prop`.
 *
 * If a falsy(inc. empty string) value is passed as `value` then that value
 * will not be applied and will be skipped.
 */
function applyStyle(el, prop, value) {
    if (!value) {
        return;
    }

    el.style[prop] = value;
}

// Extension's browser action popup UI handling to allow user to customize the
// settings of the styles. Handles the `APPLY` and `RESET` button logic.
document.addEventListener("DOMContentLoaded", async function () {
    const setFontName = (await getCurrentSetFontName()) || "";
    const setFontSize = (await getCurrentSetFontSize()) || "";
    const setLineHeight = (await getCurrentSetLineHeight()) || "";
    const applyButton = document.getElementById("apply-button");
    const resetButton = document.getElementById("reset-button");
    const fontNameInput = document.getElementById("font-name-input");
    const fontSizeInput = document.getElementById("font-size-input");
    const lineHeightInput = document.getElementById("line-height-input");

    // Set the initial value of the inputs to be the current set styles.
    fontNameInput.value = setFontName;
    fontSizeInput.value = setFontSize;
    lineHeightInput.value = setLineHeight;

    applyButton.addEventListener("click", function () {
        const font = fontNameInput.value;
        const size = fontSizeInput.value;
        const height = lineHeightInput.value;
        const fontStyles = {
            font,
            size,
            height,
        };

        // We get the details of all the tabs open and send message to all the
        // tabs with the new font styles data. All the tabs with GitHub open
        // will read this message and apply the new styles sent in the message.
        chrome.tabs.query({}, function (tabs) {
            tabs.map(function (tab) {
                chrome.tabs.sendMessage(tab.id, { data: fontStyles });
            });
        });
    });

    resetButton.addEventListener("click", function () {
        // Send message to all the tabs with data saying that we should reset the
        // styles. All the tabs with GitHub open will read this message and reset
        // styles to GitHub's default styles.
        chrome.tabs.query({}, function (tabs) {
            tabs.map(function (tab) {
                chrome.tabs.sendMessage(tab.id, { data: { reset: true } });
            });
        });
    });
});

// We listen for messages that we earlier sent when a user clicked on `APPLY` or
// `RESET` button. Based on the `request.data` of the message, we either apply
// the new styles or reset the styles.
chrome.runtime.onMessage.addListener(function (request, _sender, sendResponse) {
    const data = request.data || {};
    const shouldReset = data.reset === true;

    if (shouldReset) {
        resetStyles();
        return;
    }

    const name = data.font;
    const size = data.size;
    const height = data.height;

    setCurrentSetFontName(name);
    setCurrentSetFontSize(size);
    setCurrentSetLineHeight(height);

    const { fontFamily, fontSize, lineHeight } = genFontStyles(
        name,
        size,
        height
    );
    applyStyles(fontFamily, fontSize, lineHeight);
    sendResponse({ data, success: true });
});

// Chrome's storage API allows us to store & fetch user's recent applied styles.
// Save to chrome's storage. `data` should be an object like `{ key: value }`.
function saveToStorage(data) {
    chrome.storage.sync.set(data);
}

// Read a `value` from chromes' storage by providing the `key`. This returns a
// Promise so that we can await for the `value`. If we don't use Promise, the
// return value will always be `undefined` as chrome's storage read is async.
function getFromStorage(key) {
    return new Promise(function (resolve, _reject) {
        chrome.storage.sync.get([`${key}`], function (result) {
            const value = result[`${key}`];
            resolve(value);
        });
    });
}

// Destroy everything from the chrome's storage.
function clearStorage() {
    chrome.storage.sync.clear();
}

async function getCurrentSetFontName() {
    const currentFontName = await getFromStorage("fontName");
    return currentFontName;
}

async function getCurrentSetFontSize() {
    const currentFontSize = await getFromStorage("fontSize");
    return currentFontSize;
}

async function getCurrentSetLineHeight() {
    const currentLineHeight = await getFromStorage("lineHeight");
    return currentLineHeight;
}

function setCurrentSetFontName(name) {
    saveToStorage({ fontName: name });
    cache.fontName = name;
}

function setCurrentSetFontSize(size) {
    saveToStorage({ fontSize: size });
    cache.fontSize = size;
}

function setCurrentSetLineHeight(height) {
    saveToStorage({ lineHeight: height });
    cache.lineHeight = height;
}

// Generates correct styles by adding `px` for `size` and  adding
// `'monospace'` for `name`.
function genFontStyles(name, size, height) {
    const fontFamily = !name ? DEFAULT_FONT_FAMILY : `'${name}', 'monospace'`;
    const fontSize = !size ? DEFAULT_FONT_SIZE : `${size}px`;
    const lineHeight = !height ? DEFAULT_LINE_HEIGHT : height;

    return { fontFamily, fontSize, lineHeight };
}

async function getCurrentSetFontStyles() {
    if (cache.dirty) {
        return genFontStyles(cache.fontName, cache.fontSize, cache.lineHeight);
    }

    // Everything below here will be executed only during the first time the
    // document is loaded. After that, `cache` will always have latest font styles.

    const currentSetFontName = await getCurrentSetFontName();
    const currentSetFontSize = await getCurrentSetFontSize();
    const currentSetLineHeight = await getCurrentSetLineHeight();

    // Update the cache so that we don't make the chrome storage read calls again.
    cache.dirty = true;
    if (currentSetFontName) cache.fontName = currentSetFontName;
    if (currentSetFontSize) cache.fontSize = currentSetFontSize;
    if (currentSetLineHeight) cache.lineHeight = currentSetLineHeight;

    return getCurrentSetFontStyles();
}

// This is a cheeky solution to the issue where the code editor goes whack
// mode when we change the font styles on the code lines inside the editor.
//
// So for the time being we will not change the font styles when a user is
// editing a file on GitHub.
//
// Maybe in the future, if I get lucky( idk wtf is GitHub doing ) or someone
// in the world contributes and solves this issue where we can apply custom
// font styles to editor font without making it go insane, that would be nice.
//
// Untill then, Better GitHub won't do it's magic when user edits a file on
// GitHub. Also, who does that? xD
// Check - https://github.com/ceoshikhar/better-github/issues/6 for more info.
function isUserEditingFile() {
    function fileNameFromPath() {
        const words = window.location.pathname.split("/");
        const len = words.length;
        // The file being viewed/edited on GitHub is the last word in the URL path.
        const fileName = words[len - 1];

        return fileName;
    }

    // The word "edit" exists in the URL path if the user is editing the file.
    const wordEditInPathExists = window.location.pathname.includes("edit");
    const fileNameFromInputEl = document.querySelector(
        "input[name=filename]"
    )?.value;

    // The name of the file being edited will be same in URL path and input element on page load.
    if (fileNameFromPath() === fileNameFromInputEl && wordEditInPathExists) {
        return true;
    }

    return false;
}


scripts/build.js
/**
 * This is the build script that you can use to :
 * 1. Package the extension to better-github.zip for the specified browser.
 *    Generally for publication or distribution of the extension.
 * 2. Generate manifest.json file for the specified browser.
 *
 * Prerequisites for this script to work :
 * 1. `zip` util binary installed, run `zip --version` to check if you have it
 * or not. If you get something like `command not found: zip` then you don't
 * have it, so install it. `zip` is not needed for generating manifest.json.
 * 2. `rm -rf` should work. This is available on `linux` and `unix` machines.
 * Don't know what you need to do for it to work on `windows`. I use WSL.
 *
 * USAGE :
 * - node build <browser> [options] OR you can run scripts from `package.json`
 *
 * Valid values for `<browser>` are : "chrome" and "firefox"
 * Valid options :
 * 1. `-m` : To generate only the manifest.json and not generate package.zip
 *
 * EXAMPLE :
 * 1. Generate better-github.zip for chrome : `node build chrome`
 * 2. Generate manifest.json for firefox : `node build firefox -m`
 */

// NOTE: Update the `version` here and in `package.json` whenever a new release
// of the extension is published.
const VERSION = "1.2.0";

const fs = require("fs");
const childProcess = require("child_process");
const chalk = require("chalk");

function green(text) {
    return chalk.green(text);
}

function blue(text) {
    return chalk.blue.underline(text);
}

function result(text) {
    return chalk.black.bgYellowBright(text);
}

const chromeManifestContent = {
    manifest_version: 2,
    name: "Better Github",
    version: VERSION,
    description: "Enhance your code reading experience on GitHub",
    content_scripts: [
        {
            js: ["better-github.js"],
            matches: ["https://github.com/*", "https://gist.github.com/*"],
        },
    ],
    permissions: ["storage"],
    browser_action: {
        default_title: "Better Github",
        default_icon: "./assets/favicon.png",
        default_popup: "popup.html",
    },
    icons: {
        16: "./assets/icon16.png",
        48: "./assets/icon48.png",
        128: "./assets/icon128.png",
    },
};

// For Firefox's manifest, we just have to add one extra property: "applications"
const firefoxManifestContent = {
    ...chromeManifestContent,
    applications: {
        gecko: {
            id: "better-github@ceoshikhar.com",
            strict_min_version: "80.0",
        },
    },
};

// The first element will be 'node', the second element will be the name of the
// JS file. The next elements will be any additional command line arguments.
const args = process.argv.slice(2);
const maxArgs = 2;
const browser = args[0];
const validBrowsers = ["chrome", "firefox"];
const shouldBuildPackage = browser && args[1] === "-m" ? false : true;
const manifestName = "manifest.json";
const packageChromeName = "better-github-chrome.zip";
const packageFirefoxName = "better-github-firefox.zip";
const packageName = browserType().isChrome
    ? packageChromeName
    : packageFirefoxName;

const thingsToZip = [
    "assets/icon16.png",
    "assets/icon48.png",
    "assets/icon128.png",
    "assets/favicon.png",
    "assets/icon-no-bg.png",
    "better-github.js",
    manifestName,
    "popup.html",
    "styles.css",
];

function browserType() {
    const isChrome = browser === validBrowsers[0] ? true : false;
    const isFirefox = browser === validBrowsers[1] ? true : false;
    return { isChrome, isFirefox };
}

function makeSureArgsAreValid() {
    if (args.length > maxArgs) {
        throw new Error(`Maximum 2 arguments are allowed`);
    }

    if (!validBrowsers.includes(browser)) {
        throw new Error(
            `Invalid browser, "chrome" and "firefox" are only valid`
        );
    }

    if (shouldBuildPackage && args.length === maxArgs) {
        throw new Error(`Unexpected arguments`);
    }
}

function cleanUpOldFiles() {
    console.log(green("> Cleaning up old files if they exist"));

    const command = `rm -f manifest.json better-github-chrome.zip better-github-firefox.zip`;
    childProcess.execSync(command);
}

// Create new manifest.json and put `chromeManifestContent` if the build is
// for Chrome browser otherwise put `firefoxManifestContent` if the build is
// for Firefox browser.
function buildManifest() {
    console.log(
        green(`> Building manifest for ${browser}: `) + blue(`${manifestName}`)
    );
    if (browserType().isChrome) {
        fs.writeFileSync(
            manifestName,
            JSON.stringify(chromeManifestContent, null, 2)
        );
    } else if (browserType().isFirefox) {
        fs.writeFileSync(
            manifestName,
            JSON.stringify(firefoxManifestContent, null, 2)
        );
    }
}

function buildPackage() {
    const command = `zip ${packageName} ${thingsToZip.join(" ")}`;
    console.log(
        green(`> Building package for ${browser}: `) + blue(`${packageName}`)
    );
    childProcess.execSync(command);
}

function main() {
    const t0 = Date.now();

    makeSureArgsAreValid();
    cleanUpOldFiles();
    buildManifest();

    if (shouldBuildPackage) {
        buildPackage();
    }

    const t1 = Date.now();
    console.log(result(`> 🚀 Finished in ${t1 - t0}ms`));
}

main();


script.js
/**
 * Prerequisites for this script to work :
 * 1. `zip` util binary installed, run `zip --version` to check if you have it 
 * or not. If you get something like `command not found: zip` then you don't
 * have it, so install it. `zip` is not needed for generating manifest.json.
 * 2. `rm -rf` should work. This is available on `linux` and `unix` machines.
 * Don't know what you need to do for it to work on `windows`. I use WSL.
 * 
 * This is the script that you can use to :
 * 1. Package the extension to better-github.zip for the specified browser.
 *    Generally for publication or distribution of the extension.
 * 2. Generate manifest.json file for the specified browser. 
 * 
 * USAGE :
 * - node script <browser> [options] OR you can run scripts from `package.json`
 *  
 * Valid values for `<browser>` are : "chrome" and "firefox" 
 * Valid options :
 * 1. `-m` : To generate only the manifest.json and not generate package.zip
 * 
 * EXAMPLE :
 * 1. Generate better-github.zip for chrome : `node script chrome`
 * 2. Generate manifest.json for firefox : `node script firefox -m`
 */

const fs = require("fs");
const childProcess = require("child_process");

// NOTE: Update the `version` here and in `package.json` whenever a new release
// of the extension is published.
const chromeManifestContent = {
  "manifest_version": 2,
  "name": "Better Github",
  "version": "1.0.1",
  "description": "Enhance your code reading experience on GitHub",
  "content_scripts": [
    {
      "js": ["better-github.js"],
      "matches": ["https://github.com/*"]
    }
  ],
  "permissions": ["storage"],
  "browser_action": {
    "default_title": "Better Github",
    "default_icon": "./assets/favicon.png",
    "default_popup": "popup.html"
  },
  "icons": {
    "16": "./assets/icon16.png",
    "48": "./assets/icon48.png",
    "128": "./assets/icon128.png"
  }
};

// For Firefox's manifest, we just have to add one extra property: "applications"
const firefoxManifestContent = {
  ...chromeManifestContent,
  "applications": {
    "gecko": {
      "id": "better-github@ceoshikhar.com",
      "strict_min_version": "80.0"
    }
  }
};

// The first element will be 'node', the second element will be the name of the
// JS file. The next elements will be any additional command line arguments.
const args            = process.argv.slice(2);
const maxArgs         = 2;
const browser         = args[0];
const validBrowsers   = ['chrome', 'firefox'];
const genOnlyManifest = browser && args[1] === '-m' ? true : false;
const manifest        = 'manifest.json';
const thingsToZip     = ['assets/icon16.png', 'assets/icon48.png', 'assets/icon128.png',
                         'assets/favicon.png', 'assets/icon-no-bg.png',
                         'better-github.js', manifest, 'popup.html', 'styles.css'];
const package         = 'better-github.zip';

function browserType() {
  const isChrome  = browser === validBrowsers[0] ? true : false;
  const isFirefox = browser === validBrowsers[1] ? true : false;
  return { isChrome, isFirefox };
}

function makeSureArgsAreValid() {
  if (args.length > maxArgs) {
    throw new Error(`Maximum 2 arguments are allowed. Check docs for more info.`);
  }

  if (!validBrowsers.includes(browser)) {
    throw new Error(`Invalid browser, "chrome" and "firefox" are only valid.`);
  }

  if (!genOnlyManifest && args.length === maxArgs) {
    throw new Error(`Unexpected arguments. Check docs for more info.`);
  }
}

function maybeDeleteManifest() {
  const command = `rm -rf ${manifest}`;
  console.log(`Deleting old ${manifest} if exists`);
  childProcess.execSync(command);
}

function refreshManifest() {
  maybeDeleteManifest();

  // Create new manifest.json and put `chromeManifestContent` if the build is
  // for Chrome browser otherwise put `firefoxManifestContent` if the build is
  // for Firefox browser.
  console.log(`Generating new ${manifest} for ${browser}`);
  if (browserType().isChrome) {
    fs.writeFileSync(manifest, JSON.stringify(chromeManifestContent, null, 2));
  } else if (browserType().isFirefox) {
    fs.writeFileSync(manifest, JSON.stringify(firefoxManifestContent, null, 2));
  }
}

function maybeDeletePackage() {
  const command = `rm -rf ${package}`;
  console.log(`Deleting old ${package} if exists`);
  childProcess.execSync(command);
}

function generateNewPackage() {
  maybeDeletePackage();

  const command = `zip ${package} ${thingsToZip.join(' ')}`
  console.log("Zipping the files : ", command);
  childProcess.execSync(command);
}

function main() {
  makeSureArgsAreValid();

  if (genOnlyManifest) {
    refreshManifest();
  } else {
    refreshManifest();
    generateNewPackage();
  }
}

main();
